src/raftstore/coprocessor/split_check/table.rs

```diff
@@ -0,0 +1,488 @@
+fn check_key(status: &mut TableStatus, current_data_key: &[u8]) -> Option<Vec<u8>> {
+    if let Some(last_encoded_table_prefix) = status.last_encoded_table_prefix.take() {
+        // `before_check` found a split key.
+        return Some(keys::data_key(&last_encoded_table_prefix));
+    }
+
+    if !keys::validate_data_key(current_data_key) {
+        return None;
+    }
+    let current_encoded_key = keys::origin_key(current_data_key);
+
+    let split_key = if status.first_encoded_table_prefix.is_some() {
+        if let Ok(false) = is_same_table(
+            status.first_encoded_table_prefix.as_ref().unwrap(),
+            current_encoded_key,
+        ) {
+            // Different tables.
+            Some(current_encoded_key)
+        } else {
+            None
+        }
+    } else if is_table_key(current_encoded_key) {
+        // Now we meet the very first table key of this region.
+        Some(current_encoded_key)
+    } else {
+        None
+    };
+    if let Some(key) = split_key {
+        to_encoded_table_prefix(key).map(|k| keys::data_key(&k))
+    } else {
+        None
+    }
+}
+
+#[allow(collapsible_if)]
+fn bound_keys(db: &DB, region: &Region) -> Result<Option<(Vec<u8>, Vec<u8>)>> {
+    let start_key = keys::enc_start_key(region);
+    let end_key = keys::enc_end_key(region);
+    let mut first_key = None;
+    let mut last_key = None;
+
+    let iter_opt = IterOption::new(Some(end_key), false);
+    let mut iter = box_try!(db.new_iterator_cf(CF_WRITE, iter_opt));
+
+    // the first key
+    if iter.seek(start_key.as_slice().into()) {
+        let key = iter.key().to_vec();
+        first_key = Some(key);
+    } // else { No data in this CF }
+
+    // the last key
+    if iter.seek(SeekKey::End) {
+        let key = iter.key().to_vec();
+        last_key = Some(key);
+    } // else { No data in this CF }
+
+    match (first_key, last_key) {
+        (Some(fk), Some(lk)) => Ok(Some((fk, lk))),
+        (None, None) => Ok(None),
+        (first_key, last_key) => Err(box_err!(
+            "invalid bound, first key: {:?}, last key: {:?}",
+            first_key,
+            last_key
+        )),
+    }
+}
+
+fn to_encoded_table_prefix(encoded_key: &[u8]) -> Option<Vec<u8>> {
+    table_codec::decode_table_id(&Key::from_encoded(encoded_key.to_vec()).raw().unwrap())
+        .map(table_codec::gen_table_prefix)
+        .map(|k| Key::from_raw(&k).encoded().to_vec())
+        .ok()
+}
+
+// Encode a key like `t{i64}` will append some unecessary bytes to the output,
+// This function extracts the first 10 bytes, they are enough to find out which
+// table this key belongs to.
+const ENCODED_TABLE_TABLE_PREFIX: usize = table_codec::TABLE_PREFIX_KEY_LEN + 1;
+
+fn is_table_key(encoded_key: &[u8]) -> bool {
+    encoded_key.starts_with(table_codec::TABLE_PREFIX) &&
+        encoded_key.len() >= ENCODED_TABLE_TABLE_PREFIX
+}
+
+fn is_same_table<'a>(left_key: &'a [u8], right_key: &'a [u8]) -> StdResult<bool, &'a [u8]> {
```

Comment:
do we need the error outside? Maybe bool is enough.

Reply:
It needs to know if `left_key` and `right_key` are valid table keys, eg. `left_key` and `right_key` are both meta key, then it is indistinguishable from left meta key right table key, the former does not need to split, but the later needs.

Reply:
Where does the error get checked? I think changing it to `bool` should be enough too.